contracts/upgradeable_contracts/RTokenConnector.sol

```diff
@@ -0,0 +1,51 @@
+pragma solidity 0.4.24;
+
+import "openzeppelin-solidity/contracts/AddressUtils.sol";
+import "../interfaces/IRToken.sol";
+import "./Ownable.sol";
+import "./ERC20Bridge.sol";
+
+contract RTokenConnector is Ownable, ERC20Bridge {
+    bytes32 internal constant RTOKEN = 0xdce7cb71f7466970418465f97f1fb7d721c1661bc59b1f0caa7d1f8ce374a47e; // keccak256(abi.encodePacked("rToken"))
+
+    function initializeRToken(
+        address _rToken,
+        address[] _recipients,
+        uint32[] _proportions
+    ) external onlyOwner returns (uint256) {
+        require(AddressUtils.isContract(_rToken));
+        addressStorage[RTOKEN] = _rToken;
+        erc20token().approve(_rToken, uint256(-1));
+        return createAndChangeRTokenHat(_recipients, _proportions);
+    }
+
+    function removeRToken() external onlyOwner {
+        addressStorage[RTOKEN] = address(0);
+    }
+
+    function redeemRToken(uint256 _redeemTokens) external onlyOwner returns (bool) {
+        return _redeemRToken(_redeemTokens);
+    }
+
+    function mintRToken(uint256 _mintAmount) public returns (bool) {
+        if (rToken() == address(0)) return;
+        return IRToken(rToken()).mint(_mintAmount);
+    }
+
+    function createAndChangeRTokenHat(
+        address[] _recipients,
+        uint32[] _proportions
+    ) public onlyOwner returns (uint256) {
+        if (rToken() == address(0)) return;
+        return IRToken(rToken()).createHat(_recipients, _proportions, true);
+    }
+
+    function rToken() public view returns (address) {
+        return addressStorage[RTOKEN];
+    }
+
+    function _redeemRToken(uint256 _redeemTokens) internal returns (bool) {
+        if (rToken() == address(0)) return;
+        return IRToken(rToken()).redeem(_redeemTokens);
```

Comment:
Since we rely here on a 3rd party contracts should we make sure that all requested amount of tokens was redeemed?

Reply:
If something goes wrong then the transfer of token will fail after `redeem` (where it is used inside the contracts)

Reply:
Correct. And it will spend additional gas to call and execute the transfer. Moreover it will be hard to identify why it was failed on `transfer` by using traces. If it fails on `_redeemRTokens` we will see it in the block explorer (or by traces).

Reply:
Ok, I got your point. I will add this check